src/routes/(app)/order/[id]/payment/[paymentId]/+page.server.ts

```diff
@@ -41,20 +43,43 @@ export const actions = {
 		const formData = await request.formData();
 		const parsed = z
 			.object({
+				amount: z
+					.string()
+					.regex(/^\d+(\.\d+)?$/)
+					.nullable(),
 				method: z.enum(methods as [PaymentMethod, ...PaymentMethod[]])
 			})
 			.parse({
+				amount: formData.get('amount'),
 				method: formData.get('method')
 			});
+		let amount = 0;
+		if (parsed.amount) {
+			amount = parsePriceAmount(parsed.amount, order.currencySnapshot.main.totalPrice.currency);
+			if (amount <= 0) {
+				throw error(400, 'Invalid amount');
+			}
+		}
 		await withTransaction(async (session) => {
-			await addOrderPayment(order, parsed.method, payment.price, {
-				expiresAt: paymentMethodExpiration(parsed.method),
-				session
-			});
 			await onOrderPaymentFailed(order, payment, 'canceled', {
 				preserveOrderStatus: true,
 				session
 			});
+
+			await addOrderPayment(
+				order,
+				parsed.method,
+				amount
+					? {
+							amount,
+							currency: order.currencySnapshot.main.totalPrice.currency
+					  }
+					: payment.price,
+				{
+					expiresAt: paymentMethodExpiration(parsed.method),
+					session
+				}
```

Comment:
I think it should be `payment.price` like before?

POS users can just cancel a payment (without cancelling order, need a PR for that probably) and add back an arbitrary payment afterwards.